src/ray/core_worker/core_worker.cc

```diff
@@ -867,7 +867,11 @@ Status CoreWorker::Put(const RayObject &object,
   reference_counter_->AddOwnedObject(
       *object_id, contained_object_ids, rpc_address_, CurrentCallSite(), object.GetSize(),
       /*is_reconstructable=*/false, NodeID::FromBinary(rpc_address_.raylet_id()));
-  return Put(object, contained_object_ids, *object_id, /*pin_object=*/true);
+  auto status = Put(object, contained_object_ids, *object_id, /*pin_object=*/true);
+  if (!status.ok()) {
+    reference_counter_->RemoveOwnedObject(*object_id);
```

Comment:
I'm a bit confused by this line. How did this work before (who removed the ref if there was an error)?

Reply:
AFAICT, before this change the ref wouldn't be removed if there was a Plasma put error; the ref would remain in the reference table for the lifetime of the `CoreWorker`. I haven't confirmed this, though.

Reply:
Hmm if that's the case, I think we should leave it the way it was and not try to preemptively remove the ref. I think you're right about that; probably it's the normal ref counting codepath that removes the ref.

Reply:
I decided to trace out what would happen on a failed `ray.put()`, and it doesn't look like the normal ref counting codepath would remove the ref. If the [Plasma put](https://github.com/ray-project/ray/blob/4aef3d68360226bfaf3fa2fa8e425f4ef91873cb/python/ray/worker.py#L275-L278) fails, a non-ok status is [returned](https://github.com/ray-project/ray/blob/4aef3d68360226bfaf3fa2fa8e425f4ef91873cb/python/ray/_raylet.pyx#L1004-L1006) and a `RaySystemError` is [raised](https://github.com/ray-project/ray/blob/4aef3d68360226bfaf3fa2fa8e425f4ef91873cb/python/ray/_raylet.pyx#L152), which:

- For a [`ray.put`](https://github.com/ray-project/ray/blob/4aef3d68360226bfaf3fa2fa8e425f4ef91873cb/python/ray/worker.py#L275-L278): will abort the allocation of an object ref and therefore the [addition of a local ref](https://github.com/ray-project/ray/blob/4aef3d68360226bfaf3fa2fa8e425f4ef91873cb/python/ray/includes/object_ref.pxi#L19), and given that this will cause the object ref to never be deallocated, a [local ref will never be removed](https://github.com/ray-project/ray/blob/4aef3d68360226bfaf3fa2fa8e425f4ef91873cb/python/ray/includes/object_ref.pxi#L19), so the reference will never be deleted from the reference table.
- For [the argument of a submitted task](https://github.com/ray-project/ray/blob/4aef3d68360226bfaf3fa2fa8e425f4ef91873cb/python/ray/_raylet.pyx#L330), the task will never get submitted and it's reference count will never get incremented or decremented, so it looks like it too will never be deleted from the reference table.

Does this seem right? If so, I'd imagine that deleting the reference here is probably desired.